packages/runtime/container-runtime/src/summaryStorageCompressionHooks.ts

```diff
@@ -0,0 +1,131 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { deflate, inflate } from "pako";
+import { compress, decompress } from "lz4js";
+import { ISummaryContext } from "@fluidframework/driver-definitions";
+import { ISummaryTree, ISnapshotTree, ISummaryBlob, SummaryType } from "@fluidframework/protocol-definitions";
+import { Uint8ArrayToString } from "@fluidframework/common-utils";
+import { getBlobAtPath, listBlobPaths, replaceSummaryObject, SummaryStorageHooks } from "./summaryStorageAdapter";
+import { BlobHeaderBuilder, readBlobHeader, skipHeader, writeBlobHeader } from "./summaryBlobProtocol";
+
+export enum Algorithms {
+    NONE = 1,
+    LZ4 = 2,
+    DEFLATE = 3,
+}
+
+const ALGORITHM_KEY = "ALG";
+
+function cloneSummary(summary: ISummaryTree): ISummaryTree {
+    return JSON.parse(JSON.stringify(summary)) as ISummaryTree;
+}
+
+/**
+ * This class implements the SummaryStorageHooks which can apply various kinds of compressions
+ * to the blob payload.
+ */
+export class CompressionSummaryStorageHooks implements SummaryStorageHooks {
+    constructor(private readonly _algorithm: Algorithms) { }
+    public onPreCreateBlob(file: ArrayBufferLike): ArrayBufferLike {
+        return this.encodeBlob(file);
+    }
+    public onPostReadBlob(file: ArrayBufferLike): ArrayBufferLike {
+        return this.decodeBlob(file);
+    }
+
+    /**
+     * All paths of ISummaryTree which lead to ISummaryBlob objects are iterated.
+     * At each ISummaryBlob, content is mapped to binary array and compressed.
+     * The header is then addded which shows, which algorithm was used for encryption.
+     * The result binary is base64 encoded due to the issue with summaryTreeUploadManager#writeSummaryBlob
+     * hash assertion
+     * New ISummaryBlob is created with the new string content obtained in the above step and the
+     * old blob is replaced by this new blob in the ISummaryTree.
+     */
+    public onPreUploadSummaryWithContext(summary: ISummaryTree, context: ISummaryContext):
+    { prepSummary: ISummaryTree; prepContext: ISummaryContext; } {
+        const newSummary = cloneSummary(summary);
+        const paths: string[][] = [];
+        const currentPath: string[] = [];
+        listBlobPaths(paths, currentPath, newSummary);
+        paths.forEach((path) => {
```

Comment:
Have a look at 'recursivelyReplace()' in 'packages/dds/shared-object-base/src/serializer.ts'.

'recursivelyReplace()' descends a JSONable tree of objects, looking for a particular pattern (in this case IFluidHandles).  When it finds one, it replaces that object with a new object (the encoded or decoded FluidHandle).  If a replacement is made, it clones the parent object as the stack unwinds.

The result is a minimal copy-on-write, where only the cloned objects that those that need to be mutated (and their ancestors, which are implicitly mutated to point to the clones.)  The rest of the original tree is reused.

IIRC, the perf speedup was ~200x, although the majority of that came from avoiding JSON.parse(JSON.stringify()) to do the original clone.

Reply:
@DLehenbauer The method 'recursivelyReplace()' in 'packages/dds/shared-object-base/src/serializer.ts' looks to have a potential to be a standalone function reusable in my code either. It could be a common util method. I wonder where such method could be located. I see the @fluidframework/common-utils package but it contains only build and lib dirs.

Reply:
@DLehenbauer  Implemented in 8dfcef42de94c70fdcccb88f81b03601658f553b . It would be nice to avoid code duplication by making packages/dds/shared-object-base/src/serializer.ts#recursivelyReplace method available globally. I am not sure, where that method should be located.